File: src/rpg_assistant/ingestion/raw/stat_blocks/generic.py
from __future__ import annotations

import re

from rpg_assistant.ingestion.raw.layout import LayoutBlock, LayoutPage

from rpg_assistant.ingestion.raw.stat_blocks.types import ParsedStatBlock, StatBlockSpan

TABLE_RE = re.compile(r"(\|.+\|)|(\bAC\b|\bHP\b|\bSpeed\b)", re.IGNORECASE)
STAT_BLOCK_RE = re.compile(
    r"\b(armor class|hit points|challenge rating)\b", re.IGNORECASE
)


class GenericStatBlockProfile:
    profile_id = "generic"
    aliases: list[str] = []

    def matches_document(self, pages: list[LayoutPage]) -> bool:
        return False

    def is_false_heading(
        self,
        block: LayoutBlock,
        page_blocks: list[LayoutBlock],
        block_idx: int,
    ) -> bool:
        return block.metadata.get("stat_block_role") in {"header", "stats", "icon"}

    def detect_spans(self, pages: list[LayoutPage]) -> list[StatBlockSpan]:
        return []

    def parse_span(self, span: StatBlockSpan) -> ParsedStatBlock:
        text = "\n\n".join(block.text for block in span.blocks)
        return ParsedStatBlock(
            name="",
            raw_text=text,
            game_system=self.profile_id,
        )

    def normalize_block_text(self, text: str) -> str:
        return text

    def chunk_type_hint(self, text: str, blocks: list[LayoutBlock]) -> str | None:
        if STAT_BLOCK_RE.search(text) or TABLE_RE.search(text):
            return "stat_block"
        if len(blocks) <= 3 and max((len(b.text) for b in blocks), default=0) < 80:
            return "table"
        return None

File: packages/ingest-clj/src/rpg/ingest/stat_blocks/generic.clj
(ns rpg.ingest.stat-blocks.generic
  (:require [clojure.string :as str]
            [rpg.ingest.stat-blocks.core :as core]))

(def ^:private table-re #"(\|.+\|)|(\bAC\b|\bHP\b|\bSpeed\b)")
(def ^:private stat-block-re #"(?i)\b(armor class|hit points|challenge rating)\b")

(defmethod core/matches-document? :generic [_ _pages] false)

(defmethod core/detect-spans :generic [_ pages]
  {:pages pages :spans []})

(defmethod core/parse-span :generic [_ span]
  {:name ""
   :raw-text (->> (:blocks span) (map :text) (str/join "\n\n"))
   :game-system "generic"})

(defmethod core/false-heading? :generic [_ block _page-blocks _idx _page]
  (#{"header" "stats" "icon"} (get-in block [:metadata :stat-block-role])))

(defmethod core/normalize-block-text :generic [_ text] text)

(defmethod core/chunk-type-hint :generic [_ text blocks]
  (cond
    (re-find stat-block-re text) "stat_block"
    (re-find table-re text) "stat_block"
    (and (<= (count blocks) 3)
         (< (apply max 0 (map #(count (:text %)) blocks)) 80)) "table"
    :else nil))
